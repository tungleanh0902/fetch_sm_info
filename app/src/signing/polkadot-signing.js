export async function getSignature({ transaction }, wallet, network, curve) {
  const [{ Keyring }] = await Promise.all([
    import("@polkadot/api"),
    import("@polkadot/wasm-crypto").then(async ({ waitReady }) => {
      await waitReady()
    }),
    import("@polkadot/util-crypto").then(async ({ cryptoWaitReady }) => {
      // Wait for the promise to resolve, async WASM or `cryptoWaitReady().then(() => { ... })`
      await cryptoWaitReady()
    }),
  ])

  const keyring = new Keyring({
    ss58Format: Number(network.address_prefix),
    type: curve || network.defaultCurve,
  })
  const keypair = keyring.createFromUri(wallet.seedPhrase)
  const signedMessage = (await transaction.signAsync(keypair)).toJSON()

  return { signedMessage }
}
