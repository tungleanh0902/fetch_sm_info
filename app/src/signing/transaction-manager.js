import config from "src/../config"
import { getSigner, cancelSign, signQueue } from "./signer"
import { getGraphqlHost } from "scripts/url"
import { getFingerprint } from "scripts/fingerprint"
import { getMessage } from "./message-creator.js"
import { signWithExtension } from "scripts/extension-utils"
import gql from "graphql-tag"
import BigNumber from "bignumber.js"

const txFetchOptions = (fingerprint) => ({
  method: "POST",
  headers: {
    "Content-Type": "application/json",
    fingerprint,
    development: config.development,
  },
})

export default class TransactionManager {
  constructor(apolloClient) {
    this.apollo = apolloClient
  }

  async broadcastAPIRequest(payload) {
    const fingerprint = await getFingerprint()
    const options = {
      ...txFetchOptions(fingerprint),
      body: JSON.stringify({ payload }),
    }
    const graphqlHost = getGraphqlHost()

    return fetch(
      `${graphqlHost}/transaction/broadcast`,
      options
    ).then((result) => result.json())
  }

  async getSignQueue(submitType) {
    return signQueue(submitType)
  }

  async cancel({ userAddress, networkId }, submitType) {
    return await cancelSign(submitType, {
      address: userAddress,
      network: networkId,
    })
  }

  async getCosmosTransactionData({
    memo,
    gasEstimate,
    fee,
    senderAddress,
    network,
  }) {
    const response = await this.apollo.query({
      query: gql`
        query signingInfo($networkId: String!, $senderAddress: String!) {
          transactionMetadata(networkId: $networkId, address: $senderAddress) {
            accountNumber
            accountSequence
          }
        }
      `,
      variables: { networkId: network.id, senderAddress },
      fetchPolicy: "network-only",
    })
    const {
      data: {
        transactionMetadata: { accountNumber, accountSequence },
      },
    } = response
    const coinLookup = network.coinLookup.find(
      ({ viewDenom }) => viewDenom === fee.find(({ denom }) => denom).denom
    )
    const convertedFee = [
      {
        amount: BigNumber(fee.find(({ amount }) => amount).amount)
          .div(coinLookup.chainToViewConversionFactor)
          .toNumber(),
        denom: coinLookup.chainDenom,
      },
    ]
    return {
      accountNumber,
      accountSequence,
      chainId: network.chain_id,
      gasEstimate: String(gasEstimate),
      fee: convertedFee,
      memo,
    }
  }

  async createSignBroadcast({
    messageType,
    message,
    transactionData,
    senderAddress,
    network,
    signingType,
    password,
    polkadotAPI,
    HDPath,
    curve,
  }) {
    let broadcastableObject
    if (signingType === "extension") {
      broadcastableObject = await signWithExtension(
        messageType,
        message,
        transactionData,
        senderAddress,
        network
      )
    } else {
      broadcastableObject = await this.createAndSignLocally(
        messageType,
        message,
        transactionData,
        senderAddress,
        network,
        signingType,
        password,
        polkadotAPI,
        HDPath,
        curve
      )
    }
    return this.broadcastTransaction(
      broadcastableObject,
      messageType,
      message,
      network,
      senderAddress
    )
  }

  async createAndSignLocally(
    messageType,
    message,
    transactionData,
    senderAddress,
    network,
    signingType,
    password,
    polkadotAPI,
    HDPath,
    curve
  ) {
    const messages = await getMessage(
      network,
      messageType,
      senderAddress,
      message,
      polkadotAPI
    )
    const signer = await getSigner(
      signingType,
      {
        address: senderAddress,
        password,
        network,
        HDPath,
        curve,
      },
      config // only needed for Ledger
    )

    // TODO need to get the transaction types from db
    const { getSignableObject, getBroadcastableObject } = await import(
      `./networkMessages/${network.network_type}-transactions.js`
    )
    const signableObject = await getSignableObject(messages, transactionData)
    const signedContext = await signer(signableObject)
    const broadcastableObject = await getBroadcastableObject(
      messages,
      transactionData,
      signedContext,
      HDPath,
      curve
    )

    return broadcastableObject
  }

  async broadcastTransaction(
    broadcastableObject,
    messageType,
    message,
    network,
    senderAddress
  ) {
    const txPayload = {
      messageType,
      message,
      networkId: network.id,
      senderAddress: senderAddress,
      signedMessage: broadcastableObject,
      transaction: broadcastableObject, // to change the naming in the API as well later
    }
    const result = await this.broadcastAPIRequest(txPayload)
    if (result.success) {
      return { hash: result.hash }
    } else {
      throw Error("Broadcast was not successful: " + result.error)
    }
  }
}

// limitation of the Ledger Nano S, so we pick the top 5 rewards and inform the user.
export function getTop5RewardsValidators(rewards) {
  const rewardsPerValidatorObject = rewards.reduce((all, reward) => {
    return {
      ...all,
      [reward.validator.operatorAddress]:
        Number(reward.amount) +
        (Number(all[reward.validator.operatorAddress]) || 0),
    }
  }, {})
  const rewardsPerValidatorAddresses = Object.keys(rewardsPerValidatorObject)
  let rewardsPerValidatorArray = []
  rewardsPerValidatorAddresses.forEach((validatorAddress, index) => {
    rewardsPerValidatorArray.push({
      validator: validatorAddress,
      totalRewardAmount: Object.values(rewardsPerValidatorObject)[index],
    })
  })
  return rewardsPerValidatorArray
    .sort((a, b) => b.totalRewardAmount - a.totalRewardAmount)
    .slice(0, 5)
    .map((rewardPerValidator) => rewardPerValidator.validator)
}
