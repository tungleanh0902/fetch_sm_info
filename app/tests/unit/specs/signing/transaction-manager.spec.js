import TransactionManager from "src/signing/transaction-manager.js"
import { cancelSign, signQueue } from "src/signing/signer"

jest.mock("src/../config.js", () => ({
  default_gas_price: 2.5e-8,
  graphqlHost: "http://localhost:4000",
}))

jest.mock("scripts/fingerprint", () => ({
  getFingerprint: jest.fn(() => "iamafingerprint"),
}))

jest.mock(`src/signing/signer.js`, () => ({
  getSigner: () => () => ({
    // demoing a Cosmos signedContext here
    signature: "abcd",
    publicKey: Buffer.from("superpubkey"),
  }),
  cancelSign: jest.fn(),
  signQueue: jest.fn(),
}))

const mockFetch = jest.fn(() =>
  Promise.resolve({
    json: () => ({ success: true, hash: "abcdsuperhash" }),
  })
)

let transactionManager
describe("Transaction Manager", () => {
  // values passed in from the ActionModal component
  const defaultContext = {
    messageType: "SendTx",
    message: {
      amount: {
        amount: 2,
        denom: "ATOM",
      },
      to: ["cosmos12345"],
    },
    transactionData: {
      accountNumber: 1,
      accountSequence: 1,
      chainId: "lunie-net",
      gasEstimate: "200000",
      fee: [
        {
          amount: "10",
          denom: "uatom",
        },
      ],
      memo: "",
    },
    senderAddress: "cosmos12345",
    network: {
      id: "cosmos-hub-testnet",
      coinLookup: [
        {
          viewDenom: "ATOM",
          chainToViewConversionFactor: 1000000,
          chainDenom: "uatom",
        },
      ],
      network_type: "cosmos",
    },
    signingType: "local",
    password: "1234567890",
  }
  beforeEach(async () => {
    global.fetch = mockFetch

    transactionManager = new TransactionManager()
  })

  describe("Extension queue handling", () => {
    it("should cancel request", async () => {
      await transactionManager.cancel(
        {
          userAddress: `testaddress`,
          networkId: `testnetwork`,
        },
        `extension`
      )
      expect(cancelSign).toHaveBeenCalledWith(`extension`, {
        address: `testaddress`,
        network: `testnetwork`,
      })
    })
    it("should retrun sign queue", async () => {
      await transactionManager.getSignQueue(`extension`)
      expect(signQueue).toHaveBeenCalledWith(`extension`)
    })
  })

  describe("Create sign broadcast", () => {
    it("should create sign and broadcast a transaction for Cosmos", async () => {
      transactionManager.broadcastAPIRequest = jest
        .fn()
        .mockResolvedValue({ success: true, hash: "abcdsuperhash" })

      await transactionManager.createSignBroadcast(defaultContext)

      const broadcastableObject = {
        fee: {
          amount: [
            {
              amount: "10",
              denom: "uatom",
            },
          ],
          gas: "200000",
        },
        memo: "",
        msg: [
          {
            type: "cosmos-sdk/MsgSend",
            value: {
              amount: [
                {
                  amount: "0.000002",
                  denom: "uatom",
                },
              ],
              from_address: "cosmos12345",
              to_address: "cosmos12345",
            },
          },
        ],
        signatures: [
          {
            pub_key: {
              type: "tendermint/PubKeySecp256k1",
              value: "c3VwZXJwdWJrZXk=",
            },
            signature: "q80=",
          },
        ],
      }

      const expectArgs = {
        messageType: "SendTx",
        message: {
          amount: {
            amount: 2,
            denom: "ATOM",
          },
          to: ["cosmos12345"],
        },
        networkId: "cosmos-hub-testnet",
        senderAddress: "cosmos12345",
        signedMessage: broadcastableObject,
        transaction: broadcastableObject,
      }

      expect(transactionManager.broadcastAPIRequest).toHaveBeenCalledWith(
        expectArgs
      )
    })

    it("should handle failing broadcast", async () => {
      transactionManager.broadcastAPIRequest = jest
        .fn()
        .mockResolvedValue({ success: false, hash: 1234 })

      await expect(
        transactionManager.createSignBroadcast(defaultContext)
      ).rejects.toThrow()
    })

    it("should send broadcast request", async () => {
      await transactionManager.createSignBroadcast(defaultContext)
      expect(mockFetch.mock.calls[0]).toMatchSnapshot()
    })
  })
})
