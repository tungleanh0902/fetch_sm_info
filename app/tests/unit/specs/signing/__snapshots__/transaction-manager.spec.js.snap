// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Transaction Manager Create sign broadcast should send broadcast request 1`] = `
Array [
  "http://localhost:4000/transaction/broadcast",
  Object {
    "body": "{\\"payload\\":{\\"messageType\\":\\"SendTx\\",\\"message\\":{\\"amount\\":{\\"amount\\":2,\\"denom\\":\\"ATOM\\"},\\"to\\":[\\"cosmos12345\\"]},\\"networkId\\":\\"cosmos-hub-testnet\\",\\"senderAddress\\":\\"cosmos12345\\",\\"signedMessage\\":{\\"msg\\":[{\\"type\\":\\"cosmos-sdk/MsgSend\\",\\"value\\":{\\"from_address\\":\\"cosmos12345\\",\\"to_address\\":\\"cosmos12345\\",\\"amount\\":[{\\"amount\\":\\"0.000002\\",\\"denom\\":\\"uatom\\"}]}}],\\"fee\\":{\\"amount\\":[{\\"amount\\":\\"10\\",\\"denom\\":\\"uatom\\"}],\\"gas\\":\\"200000\\"},\\"signatures\\":[{\\"signature\\":\\"q80=\\",\\"pub_key\\":{\\"type\\":\\"tendermint/PubKeySecp256k1\\",\\"value\\":\\"c3VwZXJwdWJrZXk=\\"}}],\\"memo\\":\\"\\"},\\"transaction\\":{\\"msg\\":[{\\"type\\":\\"cosmos-sdk/MsgSend\\",\\"value\\":{\\"from_address\\":\\"cosmos12345\\",\\"to_address\\":\\"cosmos12345\\",\\"amount\\":[{\\"amount\\":\\"0.000002\\",\\"denom\\":\\"uatom\\"}]}}],\\"fee\\":{\\"amount\\":[{\\"amount\\":\\"10\\",\\"denom\\":\\"uatom\\"}],\\"gas\\":\\"200000\\"},\\"signatures\\":[{\\"signature\\":\\"q80=\\",\\"pub_key\\":{\\"type\\":\\"tendermint/PubKeySecp256k1\\",\\"value\\":\\"c3VwZXJwdWJrZXk=\\"}}],\\"memo\\":\\"\\"}}}",
    "headers": Object {
      "Content-Type": "application/json",
      "development": undefined,
      "fingerprint": "iamafingerprint",
    },
    "method": "POST",
  },
]
`;
